Updated the API design to expand the message body arguments in the API functions

britzl opened the pull request:
This PR updated the API design to expand the message body arguments in the API functions.

Previously:

```
local client = nakama.create_client(config)

local email = "super@heroes.com"
local password = "batsignal"
local body = nakama.create_api_account_email(email, password)
local session = nakama.authenticate_email(client, body)
pprint(session)
```

New:

```
local client = nakama.create_client(config)

local email = "super@heroes.com"
local password = "batsignal"
local session = nakama.authenticate_email(client, email, password)
pprint(session)
```

tomglenn commented:
Hi @britzl,

Thanks for the changes you've made so far, removing the intermediate message body creation step will hugely improve the developer experience. We have a couple more requests that will help bring the Defold SDK in line with our other SDKs.

Please move client functions to the Client object rather than have them on the Nakama object.

Instead of:
```
local client = nakama.create_client(...)
local result = nakama.add_friends(client, ...)
```

It should be:
```
local client = nakama.create_client(...)
local result = client.add_friends(...)
```

Similar to above, socket functions and events should live on the Socket object.

Instead of:
```
nakama.on_matchpresence(socket, fn)
```

It should be:
```
socket.on_match_presence(fn)
```

For reference you can take a look at the `IClient` and `ISocket` definitions in the .NET SDK.
https://github.com/heroiclabs/nakama-dotnet/blob/master/Nakama/IClient.cs
https://github.com/heroiclabs/nakama-dotnet/blob/master/Nakama/ISocket.cs

Please let me know if you have any questions or issues.

britzl commented:
Update:

* Removed the semi-colon (:) notation mentioned in the previous comment
    * `socket:send_status_update("my status")` -> `socket.send_status_update("my status")`
* Updated the example and tested against a local server
* Updated the WebSocket dependency to v 3.0.0 in game.project
   * Note: in the docs there's three dependencies list (https://heroiclabs.com/docs/nakama/client-libraries/defold/#add-defold-plugins). It is only the extension-websocket dependency that is needed

britzl commented on codegen/realtime.py:
```diff
@@ -0,0 +1,291 @@
+api = read_as_string(proto_path)
+
+
+CHANNEL_MESSAGES = [ "ChannelJoin", "ChannelLeave", "ChannelMessageSend", "ChannelMessageRemove" ]
```
These are all the messages that can be sent and that will have generated code. Is this the full list?

Also note that it's still possible to do socket.send(message) to send a message.

docsncode replied:
> Also note that it's still possible to do socket.send(message) to send a message.

Thanks, that is useful.

britzl commented on codegen/realtime.py:
```diff
@@ -0,0 +1,291 @@
+ALL_MESSAGES = CHANNEL_MESSAGES + MATCH_MESSAGES + MATCHMAKER_MESSAGES + PARTY_MESSAGES + STATUS_MESSAGES
+
+
+CHANNEL_EVENTS = [ "ChannelPresenceEvent" ]
```
List of incoming events/messages that will have listener functions generated (in the format `on_message_name`). Are these all? I think maybe also Match should be added?

tomglenn commented on README.md:
```diff
@@ -154,12 +160,10 @@ Then proceed to join a chat channel and send a message:
-local result = nakama.socket_send(socket, channel_join_message)
+local result = socket.send_channel_join_message(socket, 1, channel_id, false, false)
 
 -- send channel messages
-local channel_message_send = nakama.create_channel_message_send_message(channel_id, "Pineapple doesn't belong on a pizza!")
-local result = nakama.socket_send(socket, channel_message_send)
+local result = socket.send_channel_message_send(channel_id, "Pineapple doesn't belong on a pizza!")
```
It would be great if we could simplify these function names.

e.g. `socket.join_channel(...)` and `socket.send_channel_message(...)`

britzl replied:
Remember that this is generated code based on your API specs! For the socket and rest API functions I decided to name all functions sending message as "send_" + "message_as_snake_case".

The send_channel_message_send message is defined in your spec as ChannelMessageSend (https://github.com/heroiclabs/nakama-common/blob/master/rtapi/realtime.proto#L49) and the send_channel_join message comes form the ChannelJoin message (https://github.com/heroiclabs/nakama-common/blob/master/rtapi/realtime.proto#L41).

I guess we could skip the "send_" prefix? I'll do this!

britzl replied:
Yep, I think it works rather well without "send_" prefix on those functions

tomglenn commented on example/example.script:
```diff
@@ -47,32 +46,28 @@ function init(self)
+		local type = 1 -- 1 = room, 2 = Direct Message, 3 = Group
+		local persistence = false
+		local hidden = false
+		local result = socket.send_channel_join(target, type, persistence, hidden)
```
Would be nice if it was `socket.join_channel(...)`

tomglenn commented on nakama/nakama.lua:
```diff
@@ -4179,18 +2682,13 @@ end
-	,payload_str
-	,http_key_str
-	,callback)
+function M.rpc_func2(client, id_str, payload_str, http_key_str,callback)
```
Instead of having an `rpc_func2` that can take a payload, can we remove it and just let `rpc_func` accept a payload?

britzl replied:
Both of these are defined in apigrpc.swagger.json which means that both will get generated. I can add some special code to remove one or the other.

tomglenn commented on nakama/nakama.lua:
```diff
@@ -4475,20 +2966,14 @@ end
-	,limit_int
-	,cursor_str
-	,callback)
+function M.list_storage_objects2(client, collection_str, user_id_str, limit_int, cursor_str,callback)
```
If this is different to `list_storage_objects` it needs a clearer name and description. Currently it has the same description and signature. I would prefer to remove this entirely.

britzl replied:
Both are part of the apigrpc.swagger.json definition. I can add some code to remove one or the other?

tomglenn commented on nakama/nakama.lua:
```diff
@@ -4634,25 +3098,34 @@ end
-	,tournament_id_str
-	,body_write_tournament_record_request_tournament_record_write
-	,callback)
+function M.write_tournament_record2(client, tournament_id_str, operator, score, subscore, metadata,callback)
```
Same comment as other `"2"` functions. :)

tomglenn replied:
Apologies yes I can see they're defined that way to differentiate between POST and GET requests. Let's leave them for now, we can revisit them separately. The Unity client for example internally calls RpcFunc and RpcFunc2 depending on if the call to `RpcAsync` contains a body payload or not. We can look to do something like this in the future for Defold.

tomglenn commented on nakama/socket.lua:
```diff
@@ -0,0 +1,682 @@
+-- @param presences
+-- @param reliable
+-- @param callback
+function M.send_match_data_send(socket, match_id, op_code, data, presences, reliable, callback)
```
Can we rename this to `send_match_data`?

britzl replied:
Renamed to match_data_send which is the name of the actual protobuf message (MatchDataSend)

tomglenn commented on nakama/socket.lua:
```diff
@@ -0,0 +1,682 @@
+-- @param token
+-- @param metadata
+-- @param callback
+function M.send_match_join(socket, match_id, token, metadata, callback)
```
I think it would be great to rename some of these functions to remove the `"send"` part. e.g. here it would be `socket.join_match`. See function names in the .NET SDK for reference. https://github.com/heroiclabs/nakama-dotnet/blob/7d855ad80e5f69eb4c2841b89e9cd2dd28248385/Nakama/Socket.cs

britzl replied:
Done

tomglenn reviewed (changes requested):
This is looking great @britzl. I have a few comments to help bring it a bit further in line with the other SDKs.

tomglenn commented on nakama/socket.lua:
```diff
@@ -0,0 +1,735 @@
+--
+-- events
+--
+on_channel_presence_event
```
Should this be commented out?

britzl replied:
🤦  Doh, I tweaked the template and regenerated without doing a final check. "I only did a small change so it should work right?!"

Sorry about that.

tomglenn reviewed (commented):
Looks great @britzl, I think we're almost ready to merge. Can you have a quick look at my comments and then we'll get this pushed.

britzl commented on README.md:
```diff
@@ -168,24 +165,25 @@ local result = nakama.socket_send(socket, channel_message_send)
+* `on_status_presence_event`
+* `on_stream_data`
+* `on_channel_message`
+
```
Good point. I've updated the readme.

docsncode commented:
@tomglenn could you resolve conversations that are addressed.

docsncode commented:
@britzl @tomglenn how are we tracking a version change? With the major changes we should bump a major version.

britzl commented:
> Agreed. @britzl can you also update the `CHANGELOG` to reflect the changes.

Done. I've retroactively added previous releases to the CHANGELOG.

tomglenn reviewed (approved):
Awesome, thanks @britzl. Great work. 🚀
